Initial version

safareli opened the pull request.

safareli committed:
it compiles

safareli committed:
extract State Query Message in Time picker

safareli committed:
it works, sort of

safareli committed:
preserve order of components from format

cryogenian commented on bower.json:
```diff
@@ -0,0 +1,15 @@
+  "name": "todo",
+  "private": true,
+  "dependencies": {
+    "purescript-formatters": "git://github.com/safareli/purescript-formatters.git#interval",
```
Could you use commit hash here, not branch? Or thiese branches are wip?

safareli replied:
yes they are wip and should be merged before this is merged

cryogenian commented on src/Example.purs:
```diff
@@ -0,0 +1,73 @@
+module Example where
```
Please move example code to _example_ folder

safareli committed:
fix html

cryogenian commented on src/Example.purs:
```diff
@@ -0,0 +1,73 @@
+
+-- import Data.Map as M
+import Data.List as L
+import Debug.Trace as D
```
Please add `purescript-debug` devdependency

cryogenian commented on src/Halogen/DatePicker/Component/Time.purs:
```diff
@@ -0,0 +1,101 @@
+stateToTime ∷ State -> Time
+stateToTime s = foldr f bottom ((toUnfoldable s) :: List (Tuple F.Command Int))
+  where
+  f (Tuple F.Second val) = setSecond $ unsafeToEnum val
```
`Maybe Time` and there is no need in `unsafe` :)

cryogenian commented on src/Halogen/DatePicker/Component/Time.purs:
```diff
@@ -0,0 +1,101 @@
+      Right val -> H.put (timeToState fmt val) *> (H.raise NotifyChange)
+    pure next
+  evalPicker (GetValue next) = do
+    t <- H.get <#> stateToTime
```
`H.gets stateToTime`

cryogenian commented on src/Example.purs:
```diff
@@ -0,0 +1,73 @@
+  | HandleTimeMessage TimeIdx Time.Message a
+
+type State = Unit
+data ExampleSlot = TimeSlot TimeIdx
```
Why not just `Int`?

cryogenian commented on src/Halogen/DatePicker/Component/Types.purs:
```diff
@@ -0,0 +1,30 @@
+-- mkPicker ∷ ∀ query msg m a
+--   . PickerSpec query msg m a
+--   → Picker query msg m a
+-- mkPicker = -- TODO we need somethink like overComponent to decorate it's `eval`
```
There is such thing I suppose like `Profunctor` instance for component.

cryogenian commented on src/Halogen/DatePicker/Component/Time.purs:
```diff
@@ -0,0 +1,101 @@
+import Data.Int as Int
+
+type Message = PickerMessage
+type State = Map F.Command Int
```
It's impossible to have `HH:HH` here, right?

cryogenian commented on src/Example.purs:
```diff
@@ -0,0 +1,73 @@
+
+  setBtn ∷ TimeIdx -> String -> H.ParentHTML ExampleQuery Time.Query ExampleSlot m
+  setBtn idx str = HH.button
+    [ HE.onClick $ HE.input_ $ SetTime idx $ replaceAll (Pattern ":") (Replacement "") str ]
```
Is this `":"` replacement mandatory? Could it use string-based formatters?

cryogenian commented on src/Halogen/DatePicker/Component/Types.purs:
```diff
@@ -0,0 +1,30 @@
+  | SetValue (Either String z) a
+
+data PickerMessage
+  = NotifyChange
```
It would be great if this message carries `z` from `PickerQuery`

safareli committed:
refactor Time

safareli committed:
use Validation instead of Either in Constraint

safareli committed:
'Constraint f a' to 'Constraint a'

safareli committed:
fix typos in error messages

safareli committed:
add date picker

safareli committed:
update example

safareli committed:
fix enum generation

safareli committed:
use one numberElement

safareli committed:
use one choiseElement

safareli committed:
use date with year set to 0 for initial value

safareli committed:
add enumNumberElement

safareli committed:
move textElement

safareli committed:
add some enum types

safareli committed:
remove unused imports

safareli committed:
move constraint module

safareli committed:
small refactor

safareli committed:
fix placeholder issue

safareli committed:
Add DateTime picker and refactor Example

safareli committed:
add duration component; refactor elements

safareli committed:
fix notify staff
(actually it was not an issue at all)

safareli committed:
remove some commets

safareli committed:
fix format filename

safareli committed:
add Interval

safareli committed:
add separator to interval

safareli committed:
pass initial value as arg

safareli committed:
add error to interval setValue

safareli committed:
update number element api

safareli committed:
fix spacing

safareli committed:
update ps-validation and ps-transformers

safareli committed:
add validation in duration picker

safareli committed:
refactor duration validation

safareli committed:
refactor duration validation v2

safareli committed:
fix number input validation

safareli committed:
add build task

safareli committed:
fix mustBeMounted

safareli committed:
fix numberInput invalidity

safareli committed:
cosmetics

safareli committed:
remove duration argument from duration picker

safareli committed:
make numberComponent a Number.picker

safareli committed:
refactor number picker

safareli committed:
fix invalid style

safareli committed:
cosmetic changes

safareli committed:
rename Number.picker to Num.picker

safareli committed:
apply calidation changes to Time

safareli committed:
add Choice.picker

safareli committed:
use Choice.picker in Time.picker

safareli committed:
replace KleisliEndo with Join Star; refactor

safareli committed:
use new version of ps-enums

safareli committed:
cosmetics

safareli committed:
add validation to Date.picker

safareli committed:
add validation to DateTime picker

safareli committed:
add validation to interval

safareli committed:
fix error propagation

> if when duration is Nothing user uses fractional value incorectly
it will be parsed as valid number and input will not fail
but there combination gives invalid duration and it will stay as nothing
resulting in having valid values in input but indication of invlaidity
we need to change state to invalid if we have value for all commands

safareli committed:
value sin number input must "comute"

safareli committed:
change Meridiem to Int in Time

safareli committed:
refactor; fix warnings

safareli committed:
use Array in Format instead of List; move format modules

safareli committed:
move Internal; extract Utils

safareli committed:
remove Utils.toAlt

safareli committed:
move examples to it's own folder

safareli committed:
update bower.json, add README and travis.yml

safareli committed:
remove notes file

safareli commented:
Some of Enums could be moved to  `Data.{Formatters.DateTime,Time,Date}`

safareli commented:
In some places code might be a bit "dense", if you find such places, please mark and will refactor.

jdegoes commented:
At some point it would be nice to support actual visual date / time pickers, like Chrome's support for them or date / time picking controls on iOS. AFAICT this looks like validation on text forms with spinner controls, which is a big step up from what we have now, but not as convenient to use as the native controls.

garyb commented:
We discussed that at SlamFest actually, I thought the idea was these components would use the native controls when available and fallback as necessary / for the controls that don't exist (like intervals).

safareli commented:
Yes, fallback to native controls would be a next step.

cryogenian commented on bower.json:
```diff
@@ -0,0 +1,37 @@
+  ],
+  "dependencies": {
+    "purescript-enums": "^3.2.0",
+    "purescript-formatters": "git://github.com/safareli/purescript-formatters.git#interval",
```
It would be better if this will be commit hash

safareli replied:
purescript-formatters must be merged before this pr, so this will be changed to just version.

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+  { mkFormat :: fmtInput -> Either String fmt
+  , unformat :: fmt -> String -> Either String input
+  , picker :: fmt -> H.Component HH.HTML query Unit out m
+  , handler :: ∀z. Int -> out -> z -> Query z
```
`∀z` add whitespace please :D

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Date.purs:
```diff
@@ -0,0 +1,182 @@
+derive instance dateErrorEq :: Eq DateError
+derive instance dateErrorOrd :: Ord DateError
+derive instance dateErrorGeneric :: Generic DateError _
+instance dateErrorShow :: Show DateError where
```
Is this used anywhere? Just wondering

safareli replied:
The show instance is used in Example.

cryogenian commented on src/Halogen/DatePicker/Component/Date.purs:
```diff
@@ -0,0 +1,182 @@
+  show = genericShow
+
+type State =
+  { format :: F.Format
```
This can be changed, right? I.e. when it first created we have one format and then it can be set to another?

safareli replied:
This can't be changed using PickerQuery. will remove it from state

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/DateTime.purs:
```diff
@@ -0,0 +1,195 @@
+type Message = PickerMessage Input
+
+type State =
+  { format :: F.Format
```
This probably shouldn't be in `State` you can take it in `eval` and `render` via closures

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/DateTime.purs:
```diff
@@ -0,0 +1,195 @@
+  f a = HH.div [HP.classes [HH.ClassName "Picker-component"]] $ pure $ renderCommand a
+
+renderCommand :: ∀ m. F.Command -> HTML m
+renderCommand cmd@(F.Time fmt) = HH.slot' cpTime unit (Time.picker fmt) unit (HE.input $ Right >>> Update)
```
Why not 
```purescript
renderCommand = case _ of 
  F.Time fmt -> ...
  F.Date fmt -> ...
```

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/DateTime.purs:
```diff
@@ -0,0 +1,195 @@
+evalDateTime ∷ ∀ m . DateTimeQuery ~> DSL m
+evalDateTime (Update msg next) = do
+  s <- H.get
+  nextDateTime <- map (steper s.dateTime) $ case s.dateTime of
```
`Maybe (Either _ _)` stands for, right?
+ if not set then `Nothing`
+ else `Just` 
  * if parseable then `Right`
  * else `Left`

safareli replied:
yes

cryogenian commented on src/Halogen/DatePicker/Component/DateTime.purs:
```diff
@@ -0,0 +1,195 @@
+dateError :: DateError -> DateTimeError
+dateError x = Tuple (Just x) Nothing
+
+type StepM = Join (Star (Writer (Maybe (Tuple (Additive Int) DateTimeErrorLast)))) DateTime
```
Whoa! 😃

cryogenian replied:
Could you please explain what `Join (Star (...` means? Maybe comment?

safareli replied:
`Join (Star x)` is basicly `Endo` but for `Kleisly` (a.k.a. Star)

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,118 @@
+  {format} <- H.get
+  mbEndo <- for (unwrap format) \cmd -> do
+    num <- H.query cmd $ H.request (left <<< GetValue)
+    pure case num of
```
```purescript
pure $ (Endo <<< F.toSetter cmd) <$> join num
```

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,118 @@
+propagateChange :: ∀ m . F.Format -> Input -> DSL m Unit
+propagateChange format duration = do
+  map (mustBeMounted <<< fold) $ for (unwrap format) \cmd -> do
+    let n = (duration >>= either (const Nothing) (F.toGetter cmd <<< unIsoDuration)) :: Maybe Number
```
Is this annotation necessary?

safareli replied:
it's not ✅

cryogenian commented on src/Halogen/DatePicker/Component/Date.purs:
```diff
@@ -0,0 +1,182 @@
+derive instance dateErrorEq :: Eq DateError
+derive instance dateErrorOrd :: Ord DateError
+derive instance dateErrorGeneric :: Generic DateError _
+instance dateErrorShow :: Show DateError where
```
Why do you need generic and show instances for errors? Are they used somewhere?

safareli replied:
Generic instance is just for Show. Show is used in Example and might be useful for debugging.

cryogenian commented on src/Halogen/DatePicker/Component/Interval.purs:
```diff
@@ -0,0 +1,235 @@
+  interval -> Left $ bimap toError toError interval
+
+toError :: ∀ e a. Either (Maybe e) a -> Maybe e
+toError = either id (const Nothing)
```
This will work for any `Plus` or `Monoid`
```purescript
leftOrEmpty :: forall e m a. Plus m => Either (m a) a -> m a
leftOrEmpty = either id $ const empty
```

```purescript
leftOrEmpty:: forall m a b. Monoid m => Bifoldable b => b m a -> m 
leftOrEmpty = bifoldMap id $ const mempty
```
(Isn't blocker, just note)

safareli replied:
✅ 

``` purescript
asLeft ∷ ∀ e a f. Alternative f => Either e a → f e
asLeft = either pure (const empty)

toError = asLeft >>> join
```

related
https://github.com/purescript/purescript-either/issues/31

cryogenian commented on src/Halogen/DatePicker/Component/Interval.purs:
```diff
@@ -0,0 +1,235 @@
+evalPicker (Base (SetValue interval next)) = do
+  {format} <- H.get
+  res <- case viewInterval format interval of
+    Just (StartEnd a b) -> for_ a (setDateTime false) *> for_ b (setDateTime true) $> Nothing
```
Could you please use do-notation here in such places?

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Time.purs:
```diff
@@ -0,0 +1,170 @@
+      Just (Right time) -> pure $ update time
+      _  -> buildTime
+  H.modify _{ time = nextTime }
+  when (nextTime /= s.time) $ H.raise (NotifyChange nextTime)
```
`unless`

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Time.purs:
```diff
@@ -0,0 +1,170 @@
+  mbKleisliEndo <- for (sort $ unwrap format) mkKleisli
+  pure $ map fold (sequence mbKleisliEndo) >>= \(Join (Star f)) -> f bottom
+  where
+  mkKleisli (F.Placeholder _) = pure $ Just $ mempty
```
Oh! That's why `Join (Star ...` here, right? 
Could you add type annotation to `mkKleisli` and maybe type synonim for `Join (Star ...` thing?

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Format/DateTime.purs:
```diff
@@ -0,0 +1,126 @@
+go :: List Command -> StateT FDT.Formatter (Either String) (List Command)
+go currFmt = get >>= \a -> case (takeDate a <|> takeTime a) of
+  Just (Left err)  -> lift $ Left err
+  Just (Right (Tuple restRes restFmt)) -> let res = put restFmt *> pure (currFmt <> restRes) in
```
Please break this to 5-6 lines :wink: e.g. 
```purescript
let 
  res = do 
    put restFmt
    pure $ currFmt <> restRes
in 
  if null restFmt 
    then res
    else res >>= go
```

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Format/DateTime.purs:
```diff
@@ -0,0 +1,126 @@
+  where
+  toDTCommand (Date inFmt) = DateF.toDateTimeFormatter inFmt
+  toDTCommand (Time inFmt) = TimeF.toDateTimeFormatter inFmt
+  -- toDTCommand (Placeholder str) = FDT.Placeholder str
```
commented code. If `Placeholder` is temporarily commented, it's enough to have a comment there I think

safareli replied:
✅

cryogenian replied:
? Comment is still here

safareli replied:
✅ ✅

cryogenian commented on src/Halogen/DatePicker/Format/DateTime.purs:
```diff
@@ -0,0 +1,126 @@
+unformat ∷ Format -> String -> Either String DateTime
+unformat fmt str = FDT.unformat (toDateTimeFormatter fmt) str
+
+format ∷ Format -> DateTime -> String
```
Formatting is different in different files: `∷` and `::`, arrows, forall etc.

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Format/Duration.purs:
```diff
@@ -0,0 +1,65 @@
+instance formatShow :: Show Format where
+  show = genericShow
+
+
```
Why so many blank lines?

safareli replied:
✅  now i use max 2 empty lines

cryogenian commented on src/Halogen/DatePicker/Format/Duration.purs:
```diff
@@ -0,0 +1,65 @@
+toGetter cmd = unwrap >>> lookup cmd
+
+mkFormat :: ∀ f. Foldable f => f Command -> Either String Format
+mkFormat cmds = if (errs /= [])
```
`not $ null errs`?

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Format/Time.purs:
```diff
@@ -0,0 +1,167 @@
+
+
+toSetter :: Command -> Int -> Time -> Maybe Time
+toSetter Hours24 n t = toEnum n <#> ( _ `setHour` t)
```
I think following is easier to read
```purescript
toSetter cmd n t = case cmd of 
  Hours24 n t -> 
    toEnum n <#> (_ `setHour` t)
  Hours12 n t -> 
    toEnum n >>= (_ `setHour12` t)
  ...
```

Same for `toGetter` etc.

safareli replied:
made this change but left right hand side of -> on same line as it's too small
✅

cryogenian commented on src/Halogen/DatePicker/Format/Time.purs:
```diff
@@ -0,0 +1,167 @@
+fromDateTimeFormatter :: FDT.Formatter -> Either String Format
+fromDateTimeFormatter fmt = do
+  let errs = C.runConstraint formatConstraint fmt
+  when (errs /= []) $ Left $ joinWith "; " errs
```
`unless (null errs)`

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Internal/Choice.purs:
```diff
@@ -0,0 +1,149 @@
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties as HP
+
+
```
Why 3 blank lines?

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Internal/Constraint.purs:
```diff
@@ -0,0 +1,69 @@
+runConstraint f a = unV id (const []) $ f a
+
+allowedValues :: ∀ g a. Foldable g => (a -> String) -> Array (EqPred a) -> Constraint (g a)
+allowedValues showVal as as' = for_ as' \a ->
```
```purescript
for_ as' \a -> unless (matchesAny a as) $ invalid blablabla
```

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Internal/Constraint.purs:
```diff
@@ -0,0 +1,69 @@
+  else invalid ["Contains value: " <> (showVal a) <> " which is not allowed"]
+
+notEmpty :: ∀ f a. Foldable f => Constraint (f a)
+notEmpty as = if null as then invalid ["Input must contain values"] else pure unit
```
`when (null as)`

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Internal/Constraint.purs:
```diff
@@ -0,0 +1,69 @@
+allowNoneOrAll :: ∀ f a. Foldable f => Array (EqPred a) -> Constraint (f a)
+allowNoneOrAll as = f <<< usageCount as
+  where
+  f c | c /= 0 && c /= length as = invalid ["Usage count (" <> show c <> ") for allowed elements (" <> (show as) <> ") must be 0 or all"]
```
Break line, please

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Date.purs:
```diff
@@ -0,0 +1,182 @@
+import Halogen as H
+import Halogen.Component.ChildPath as CP
+import Halogen.Datapicker.Format.Date as F
+import Halogen.Datapicker.Internal.Choice as Choice
```
Could you provide `Halogen.Datepicker.Internal` reexporting all package content?

safareli replied:
Why? There would be conflict in Num and Choice pickers

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+import Data.Functor.Coproduct.Nested as Coproduct
+import Data.Interval as I
+import Halogen as H
+import Halogen.Component.ChildPath as CP
```
Could you please sort imports alphabetically?

safareli replied:
✅

cryogenian reviewed (changes requested):
I'll review it once again next morning. Currently 

+ It should be `Datepicker` not `Datapicker`
+ In general `if _ then pure unit else _` and `if _ then _ else pure unit` are `unless` and `when`
+ There are couple formatting issues: double colons, foralls, blank lines etc. It seems to me that `case`s are easier to read than function clauses, idk maybe it's my personaly preferences.
+ I'm not sure that I understood correctly, but aren't `Show` instances used for rendering? 
+ `format` argument shouldn't go to state, you can thread it to your eval and render funcs (`Reader` if you dislike funcs)
+ Adding comments especially for nested `Maybe (Either (Maybe ` and `Bitraverse` things is good idea probably. 

I really like the code though :100:

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+  | HandleMessage MessagePayload a
+
+data SetPayload
+  = SetTime Int (Maybe Time)
```
This `Int` is `TimeIdx`, right? Maybe you could switch to type synonims here too, or use `Int` as slot? 
I'm not sure that it worth to try using newtypes here, but :)

safareli replied:
✅

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+  eval ∷ Query ~> DSL m
+  eval (Set payload next) = do
+    map mustBeMounted $ case payload of
+      SetTime     idx val -> H.query' timeConfig.cp     idx $ H.request $ left <<< Base <<< (SetValue $ map Right val)
```
Could you please split these lines? Do you have strong preferences in alignment `H.request`? I mean this is probably would be easier to read
```purescript
  SetTime idx val ->
    H.query' timeConfig.cp idx
      $ H.request 
      $ left <<< Base <<< (SetValue $ map Right val)
```

safareli replied:
✅

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+        pure $ void res -- <#> (\error ->  D.traceAny {message:"can't update interval", error})
+    pure next
+  eval (HandleMessage payload next) = do
+    case payload of
```
```purescript
H.modify \s -> case payload of 
  MsgTime idx (NotifyChange val) -> s{ times = insert idx }
  ... -> ...
```

safareli replied:
✅

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+    case payload of
+      MsgTime     idx (NotifyChange val) -> H.modify \s -> s {times     = insert idx (show val) s.times}
+      MsgDate     idx (NotifyChange val) -> H.modify \s -> s {dates     = insert idx (show val) s.dates}
+      MsgDateTime idx (NotifyChange val) -> H.modify \s -> s {dateTimes = insert idx (show val) s.dateTimes}
```
Frankly I think that such alignment is kinda trick showing values as similar, but in fact you can use newlines in `case` clauses and abstract things like `insert idx (show val)` and it _should_ become useless 

(That's my personal opinion, if you have strong preferences, feel free to leave it as is)

safareli replied:
✅

didn't get what you mean by "abstract things like insert idx (show val) and it" tho

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+  , cp :: CP.ChildPath query ChildQuery Int Slot
+  }
+
+renderExample :: ∀ fmtInput input fmt query out m
```
Could you please put `:: forall fmtInput input fmt query out m` to newline?

safareli replied:
✅

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,327 @@
+  unEither :: Either String (Array (HTML m)) -> Array (HTML m)
+  unEither = either (HH.text >>> pure >>> HH.div_ >>> pure) id
+
+timeConfig :: ∀ m. ExampleConfig String Time TimeF.Format Time.Query Time.Message m
```
6 type params are a bit hard to visually parse. Maybe newlines or type synonims could help.

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Date.purs:
```diff
@@ -0,0 +1,182 @@
+
+renderCommand :: ∀ m. F.Command -> HTML m
+renderCommand cmd = case cmd of
+  F.Placeholder str     -> textElement { text: str}
```
Newlines please

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Date.purs:
```diff
@@ -0,0 +1,182 @@
+  mbKleisliEndo <- for (sort $ unwrap format) $ commandCata
+    { text: \cmd -> pure $ Just $ mempty
+    , enum: \cmd -> do
+      num <- H.query' cpNum cmd $ H.request (left <<< GetValue)
```
Please add 2 spaces here, do-block and record are visually mixed

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/DateTime.purs:
```diff
@@ -0,0 +1,195 @@
+  applyTime :: PickerValue TimeError Time -> StepM
+  applyTime val = Join $ Star $ \dt -> case val of
+    Just (Right time) -> pure $ setTimeDt time dt
+    Just (Left err) -> tell (Just $ Tuple (Additive 1) $ bimap Last Last $ timeError err) *> pure dt
```
Newlines please. Compare
```purescript
tell (Just $ Tuple (Additive 1) $ bimap Last Last $ timeError err)
```
and 
```purescript
  tell 
  $ Just 
  $ Tuple (Additive 1)
  $ bimap Last Last 
  $ timeError err
```

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/DateTime.purs:
```diff
@@ -0,0 +1,195 @@
+setDateDt x dt = modifyDate (const x) dt
+
+
+stepsToFunc :: ∀ f. Foldable f => Int -> f StepM -> DateTime -> Either (Tuple Boolean DateTimeError) DateTime
```
Split type annotation to multiple lines, please

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,118 @@
+evalDuration ∷ ∀ m . DurationQuery ~> DSL m
+evalDuration (UpdateCommand cmd val next) = do
+  s <- H.get
+  nextDuration <- map (steper' s.duration InvalidIsoDuration) $ case s.duration of
```
Unneeded `$`

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,118 @@
+evalPicker (Base (SetValue duration next)) = do
+  {format} <- H.get
+  propagateChange format duration
+  H.modify \s -> s{duration = duration}
```
`H.modify _{ duration = duration }`

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,118 @@
+  propagateChange format duration
+  H.modify \s -> s{duration = duration}
+  pure $ next unit
+evalPicker (Base (GetValue next)) = H.gets _.duration <#> next
```
I think this should be `k` or `cont` or `reply`

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Interval.purs:
```diff
@@ -0,0 +1,235 @@
+  pure $ next res
+evalPicker (Base (GetValue next)) = H.gets _.interval <#> next
+
+viewInterval :: F.Format -> Input -> Maybe (Interval (Maybe Duration.Input) (Maybe DateTime.Input))
```
Split type annotation to multiple lines please

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Interval.purs:
```diff
@@ -0,0 +1,235 @@
+  _, Nothing -> Just $ bimap (const $ Just Nothing) (const $ Just Nothing) format
+  _ , _ -> Nothing
+  where
+  mapedInput :: Input -> Maybe (Interval (Maybe Duration.Input) (Maybe DateTime.Input))
```
Maybe maybe maybe :D type annotation may help here

safareli replied:
✅

safareli committed:
use when/unless

safareli committed:
refactor Constraint Errors

safareli committed:
max 2 empty lines

safareli committed:
fix foralls

safareli committed:
fix double colons

safareli committed:
replace Datapicker with Datepicker

safareli committed:
add space after ∀

safareli committed:
sort imports

safareli committed:
fix some case formating

safareli committed:
`when (x /= y)` to `unless (x == y)`

safareli committed:
fix arrow

safareli committed:
use more case of

safareli committed:
int to index type in example

safareli committed:
remove `$` before `case`

safareli commented:
fixed some of small issues will continue tomorrow

safareli committed:
use asRight

safareli committed:
use join in buildDuration

safareli committed:
add asLeft

safareli committed:
use do in interval eval

safareli committed:
remove format from state

safareli committed:
add moveStateTo

safareli committed:
use transitionState in eval

safareli committed:
replace => with ⇒

safareli committed:
replace <- with ←

safareli committed:
make lines shorter; some cosmetic changes

safareli committed:
cosmetic refactoring

safareli committed:
utf fix

safareli commented:
I think it's ready for next review.
I will try to make `buildxxx1 staff a bit simpler tomorrow

safareli committed:
fix replace errors

cryogenian commented on example/src/Main.purs:
```diff
@@ -0,0 +1,404 @@
+      , I.second 124.0
+      ]
+
+  -- TODO refactor this
```
Still `TODO`?

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,105 @@
+
+type Slot = F.Command
+
+
```
What's the semantic difference between `Slot` synonim and `HTML|DSL` synonims?

safareli replied:
I have `type Slot` in all components so I have this one here too, for convenience.
(I think I don't understand you question tho)

cryogenian replied:
Oh! I mean that during reading the code blank lines usually mean that things below are slightly different with previous. That's why I asked :)

safareli replied:
oh yes. will try to make spaces more consistant

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,105 @@
+    (HE.input $ \(NotifyChange n) → UpdateCommand cmd n)]
+
+getComponent ∷ F.Command → IsoDuration → Number
+getComponent cmd d = maybe 0.0 id $ F.toGetter cmd (unIsoDuration d)
```
`maybe a id === fromMaybe a`

safareli replied:
✅

cryogenian commented on src/Halogen/DatePicker/Component/Duration.purs:
```diff
@@ -0,0 +1,105 @@
+propagateChange ∷ ∀ m . F.Format → State → DSL m Unit
+propagateChange format duration = do
+  map (mustBeMounted <<< fold) $ for (unwrap format) \cmd → do
+    let n = duration >>= asRight >>= (F.toGetter cmd <<< unIsoDuration)
```
Are these parens needed?

safareli replied:
it's not ✅

cryogenian commented on src/Halogen/DatePicker/Format/DateTime.purs:
```diff
@@ -0,0 +1,142 @@
+    resFmt ← evalStateT s fmt
+    let errs = C.runConstraint formatConstraint resFmt
+    when (errs /= []) $ Left $ joinWith "; " errs
+    pure $ Format $ fromFoldable resFmt --- TODO make sure List used here is fine
```
`TODO`

safareli replied:
✅  List is fine as we just have two commands

cryogenian commented on src/Halogen/DatePicker/Internal/Utils.purs:
```diff
@@ -0,0 +1,69 @@
+  where
+  classes = [ClassName "Picker"] <> (guard (isInvalid val) $> ClassName "Picker--invalid")
+
+componentProps ∷ forall r z. Array (HP.IProp ( "class" ∷ String | z ) r )
```
This module has `forall`s and `∀` :)

safareli replied:
✅

safareli committed:
refactor build functions

safareli committed:
remove perens

safareli committed:
remove comments

safareli committed:
fix forall

safareli committed:
remove comments

safareli committed:
refactor
- don't use unsafePartialBecause, but use halt
- add common request/actions to .Types

safareli committed:
reorder component types in more consistnat way

order:

- State
- Message
- Query
- ChildSlot&Slot
- ChildPath
- HTML&DSL

cryogenian reviewed (approved):
:+1: Feel free to merge this after deps are updated (formatters, datetime)

safareli committed:
update ps-datetime

safareli committed:
add placeholders

safareli committed:
refactor html query maping

safareli committed:
update purescript-datetime version

safareli committed:
update ps-foramtters

safareli committed:
add length classes to input

safareli committed:
add configuration for altering styles

cryogenian commented on src/Halogen/DatePicker/Internal/Elements.purs:
```diff
@@ -1,9 +1,77 @@
+toNumConf (Config {input, inputInvalid, inputLength}) ({title, placeholder, range}) =
+  {title, placeholder, range, root: input, rootInvalid: inputInvalid, rootLength: inputLength }
+
+renderCommandNum :: forall m slot cmd childQuery parentQuery valIn valOut.
```
Inconsistent `foralls` 😉

cryogenian reviewed (approved):
Found one formatting thing, I don't think it should prevent merging.

safareli committed:
cleen up type signature of Elements.*

safareli committed:
add choiceEmptyTitle

safareli committed:
add Newtype instance for Config

safareli commented:
Ready to merge

beckyconning commented on example/src/Main.purs:
```diff
@@ -194,7 +195,7 @@ example =
           , DurationF.Second
           ]
-          "YYYY:MM:DD-HH:mm"
+          "YYYY:MMM:DD-HH:mm"
```
This change doesn't seem related to the commit title. Why was it made?

beckyconning replied:
Oh I see you were testing the empty title. Makes sense!

safareli replied:
yap

beckyconning commented:
This is amazing @safareli. From my perspective you've solved a huge problem with web interfaces in a very elegant and useful way. This is an amazing piece of UI technology and I can't wait to see it being put to work by our customers and as a part of other pieces of software.

safareli commented:
Thanks @beckyconning 🎉
